File: le1/plp/expressions1/expression/ExpMenos.java
package le1.plp.expressions1.expression;

import le1.plp.expressions1.util.Tipo;

/**
* Um objeto desta classe representa uma Expressao de menos unario.
*/

public class ExpMenos extends ExpUnaria{

	/**
	 * Controi uma Expressao de menos unario com expressao especificada
	 *
	 * @param exp Expressao cuja avaliacao resulta <code>ValorInteiro</code>.
	 */
	public ExpMenos(Expressao exp){
		super(exp, "-");
	}

	/**
	 * Retorna o valor da Expressao de menos unario
	 */
	public Valor avaliar(){
		return new ValorInteiro(- ((ValorInteiro)getExp().avaliar()).valor());
	}

	/**
	 * Realiza a verificacao de tipos desta expressao.
	 *
	 * @return <code>true</code> se os tipos da expressao sao validos;
	 *          <code>false</code> caso contrario.
	 */
	protected boolean checaTipoElementoTerminal() {
		return (getExp().getTipo().eInteiro());
	}

	/**
	 * Retorna os tipos possiveis desta expressao.
	 *
	 * @return os tipos possiveis desta expressao.
	 */
	public Tipo getTipo() {
		return Tipo.TIPO_INTEIRO;
	}
}

File: Le1 - Python/le1/expression/exp_menos.py
from __future__ import annotations

from le1.expression.exp_unaria import ExpUnaria
from le1.expression.expressao import Expressao
from le1.expression.valor_inteiro import ValorInteiro
from le1.util.tipo import Tipo


class ExpMenos(ExpUnaria):
    def __init__(self, exp: Expressao) -> None:
        super().__init__(exp, "-")

    def avaliar(self) -> ValorInteiro:
        valor = self.getExp().avaliar()

        if not isinstance(valor, ValorInteiro):
            raise TypeError("ExpMenos espera ValorInteiro")

        return ValorInteiro(-valor.valor())

    def checaTipoElementoTerminal(self) -> bool:
        return self.getExp().getTipo().eInteiro()

    def getTipo(self) -> Tipo:
        return Tipo.TIPO_INTEIRO